File: Leetcode/PY/Leetcode519-随机翻转矩阵.py
class Solution:

    def __init__(self, m: int, n: int):
        self._m = m
        self._n = n
        self.k = m * n
        self.mp = defaultdict(int)

    def flip(self) -> List[int]:
        idx = random.randint(0,self.k - 1)
        real = idx
        if idx in self.mp: real = self.mp[idx]
        if (self.k - 1) in self.mp:
            self.mp[idx] = self.mp[self.k - 1]
            del self.mp[self.k - 1]
        else:
            self.mp[idx] = self.k - 1
        self.k -= 1
        return [real // self._n,real % self._n]


    def reset(self) -> None:
        self.k = self._m * self._n
        self.mp = defaultdict()

File: Leetcode/JS/Leetcode519-随机翻转矩阵.js
/**
 * @param {number} m
 * @param {number} n
 */
 var Solution = function(m, n) {
    this.m = m,this.n = n;
    this.k = m * n;
    this.mp = new Map();
};

/**
 * @return {number[]}
 */
Solution.prototype.flip = function() {
    let idx = Math.floor(Math.random() * (this.k - 1));
    let real = idx;
    if(this.mp.has(idx)) real = this.mp.get(idx);
    if(this.mp.has(this.k - 1)) {
        this.mp.set(idx,this.mp.get(this.k - 1));
        this.mp.delete(this.k - 1);
    }else{
        this.mp.set(idx,this.k - 1);
    }
    this.k--;
    return [Math.floor(real / this.n),real % this.n];
};  

/**
 * @return {void}
 */
Solution.prototype.reset = function() {
    this.mp = new Map();
    this.k = this.m * this.n;
};  
